mergify_engine/merge_train.py

```diff
@@ -0,0 +1,422 @@
+# -*- encoding: utf-8 -*-
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License. You may obtain
+# a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import dataclasses
+import json
+import logging
+import typing
+from urllib import parse
+
+import daiquiri
+import redis
+
+from mergify_engine import check_api
+from mergify_engine import constants
+from mergify_engine import context
+from mergify_engine import github_events
+from mergify_engine import github_types
+from mergify_engine import rules
+from mergify_engine import subscription
+from mergify_engine import utils
+from mergify_engine.clients import github
+from mergify_engine.clients import http
+
+
+WaitingPull = typing.NamedTuple("WaitingPull", [("pull", int), ("queue_name", str)])
+
+
+@dataclasses.dataclass
+class TrainCar:
+    train: "Train" = dataclasses.field(repr=False)
+    pull: int
+    queue_name: str
+    parents: typing.List[int]
+    initial_current_base_sha: github_types.SHAType
+    current_base_sha: github_types.SHAType
+    tmp_pull_number: typing.Optional[int] = dataclasses.field(default=None)
+
+    class Serialized(typing.TypedDict):
+        pull: int
+        queue_name: str
+        parents: typing.List[int]
+        initial_current_base_sha: github_types.SHAType
+        current_base_sha: github_types.SHAType
+        tmp_pull_number: typing.Optional[int]
+
+    def serialized(self) -> "TrainCar.Serialized":
+        return self.Serialized(
+            pull=self.pull,
+            queue_name=self.queue_name,
+            parents=self.parents,
+            initial_current_base_sha=self.initial_current_base_sha,
+            current_base_sha=self.current_base_sha,
+            tmp_pull_number=self.tmp_pull_number,
+        )
+
+    @classmethod
+    def deserialized(cls, train: "Train", data: "TrainCar.Serialized") -> "TrainCar":
+        return cls(train, **data)
+
+    def create_pull(self) -> None:
```

Comment:
I'm a bit worried we build things as non-async but I guess we don't have much choice?

Reply:
async in sense of asyncio or async in sense of we do that while we run the queue action  ?

Reply:
In the sense of asyncio